src/server.c

```diff
@@ -98,6 +98,11 @@ int isReadyToShutdown(void);
 int finishShutdown(void);
 const char *replstateToString(int replstate);
 
+/*============================ AE event modification functions ============================ */
+int aeCreateFileEvent(aeEventLoop *eventLoop, int fd, int mask, aeFileProc *proc, void *clientData);
+void aeDeleteFileEvent(aeEventLoop *eventLoop, int fd, int mask);
+int aeResizeSetSize(aeEventLoop *eventLoop, int setsize);
```

Comment:
We can avoid duplication of these declarations and do this?

```suggestion
#include "ae.h"
```

Reply:
Since `AE_EXCLUDE_EVENT_MODIFICATION` is defined in `connection.h` to ensure we don't directly access the modify-events functions, and since `connection.h` is included in `server.h`, including `ae.h` alone won't be sufficient.

Reply:
Yes, it will be complicated. (Maybe it works if server.c includes ae.h before it includes server.h, but it's a too strange dependency between the files.)

The conditional prototypes in ae.h makes things a bit complicated. Let's think about some way to avoid it.

One idea from the top of my head: Can we move these from ae.h to a separate ae_modify.h or something like that?

Do we actually need to hide these command prototypes with the ifdef? I understand it is to protect against concurrent usage from threads, but we have many other functions too that could potentially be accessed from threads, but we just know they're not used by threads.

Reply:
I'll check the ` ae_modify.h` option,
Regarding your question, the event loop is accessed by threads during epoll offload, so we must acquire the mutex before accessing the eventloop struct from the **_main thread_**. As you mentioned in PR #477, we modify the eventloop struct without taking the mutex. Without the ifdef guard, there's no guarantee against modifying server.el without mutex protection. This differs from functions where we know threads won't access, as changes to IO thread code are made with awareness of potential race conditions with the main thread.